dealiasing and annotations

paulp opened the pull request:
That work was intersecting too much with another pending pull request so they met in the middle. This addresses I hope all the many comments left on #1804, and then some. Review by @xeno-by.

paulp committed:
Renamed methods to be less ambiguous in intent.

isNamed    => isNamedArg
isIdentity => allArgsArePositional
nameOf     => nameOfNamedArg

Moved mkNamedArg into TreeGen.

paulp committed:
SI-6083, misleading annotation error message.

When an implicit conversion causes an apparently constant
argument not to be constant, show the conversion, not the
constant.

paulp committed:
SI-5182, no position on annotation error.

Now there's a position on the synthetic "value" Ident.

paulp committed:
SI-2577, SI-6860: annotation type inference.

This is less than ideal:

  scala> class Bippy[T] extends annotation.StaticAnnotation
  defined class Bippy

  scala> def f: Int @Bippy = 5
  f: Int @Bippy[T]

Turns out we can infer such types. Now it says:

  scala> def f: Int @Bippy = 5
  f: Int @Bippy[Nothing]

This should put to rest many an issue with parameterized
annotations.

paulp committed:
Removing superfluous method parameters.

Adding a method parameter with a default argument which
deviates from the default at a single call site is a pretty
questionable tradeoff. Even more so when done twice.

Like sands through the hourglass, so go the parameters
of our method.

paulp committed:
Added test for untested nested annotation restriction.

paulp committed:
Cleaning up dummy applied types and friends.

Incorporates feedback from dealias/widen PR.

paulp committed:
Add PolyType to Infer#normalize.

It arises when inferring the type of an overloaded call:

  def g(s: String): String = s
  def f: String = ???
  def f[C](c: C): String = g(f)

Also refined warning when isHKSubType is called with
arguments which very likely were never meant to be compared.

paulp committed:
A very interesting checkfile update.

Surely signfiicant, but I haven't determined in what way.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1407/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1407/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2130/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2130/

paulp commented:
Oh cool, the sbt test caught me breaking sbt.

paulp committed:
Put back a method which sbt is using.

Good catch, ant target sbt.compile.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2142/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2142/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1416/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1416/

scala-jenkins commented:
Started jenkins job JenkinsJob(pr-rangepos) at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1487/

paulp commented:
Rebased. Hey @xeno-by, please come and at least rubber stamp this, I poured a lot of work into it on the basis of your many comments, don't let your creation die of neglect.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2349/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1618/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1618/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

adriaanm commented:
check file probably needs updating:

```
  [partest] testing: [...]/files/neg/t5182.scala                                  [FAILED]
  [partest] t5182.scala:2: error: unknown annotation argument name: qwe
  [partest]   @java.lang.Deprecated(qwe = "wer") def ok(q:Int) = 1
  [partest]                             ^
  [partest] t5182.scala:3: error: classfile annotation arguments have to be supplied as named arguments
  [partest]   @java.lang.Deprecated("wer") def whereAmI(q:Int) = 1
  [partest]                         ^
  [partest] two errors found
```

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Checkable.scala:
```diff
@@ -62,6 +62,9 @@ trait Checkable {
       val tps1 = (from baseType bc).typeArgs
       val tps2 = (tvarType baseType bc).typeArgs
+      if (tps1.size != tps2.size)
+        devWarning(s"Unequal type arg lists in propagateKnownTypes($from, $to): ($tps1, $tps2)")
```
How about "unequally sized"?

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2349/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3540,7 +3546,7 @@ trait Typers extends Adaptations with Tags {
             // sometimes does. The problem is that "self" ident's within
             // annot.constr will retain the old symbol from the previous typing.
             val func     = Function(funcparm :: Nil, ann.duplicate)
-            val funcType = appliedType(FunctionClass(1), selfsym.info, annClass.tpe_*)
```
Why was it `tpe_*` here, but `tpe` on the line 3529?

xeno-by replied:
Using `annType` here is necessary to propagate the info about the inferred type, rght?

paulp replied:
If you look at the implementation of def tpe, it's tpe__. tpe__ is how you say you've read my comments not to use .tpe and are taking responsibility for the possibility the resulting type contains dummies. And yes, annType is the actual inferred type, rather than the dummy-carrying fake type you would get from annClass.tpe_*.

lrytz replied:
should we deprecate tpe?

paulp replied:
I think so; my only problem with that is that it doesn't go far enough. But it's a start.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3450,44 +3461,40 @@ trait Typers extends Adaptations with Tags {
-        extract(ann, List())
-      }
+      val treeInfo.Applied(fun0, targs, argss) = ann
+      val typedFun0 = typed(fun0, mode.forFunMode, WildcardType)
```
Previously typed(fun0) wasn't called if isErroneous happened to be true. Isn't the current approach going to sometimes cause spurious errors?

paulp replied:
I don't know, I don't understand any error code. Apparently it doesn't cause any with our current test coverage. If it matters I can change it.

xeno-by replied:
@hubertp, could you please comment on this matter?

paulp replied:
In the interests of seeing this merged in this lifetime I hacked it to make it like it was before.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3450,44 +3461,40 @@ trait Typers extends Adaptations with Tags {
-            case Select(New(tpt), nme.CONSTRUCTOR) =>
-              (fun, outerArgss)
-            case _ =>
-              reportAnnotationError(UnexpectedTreeAnnotation(fun))
```
Looks like UnexpectedTreeAnnotation is no longer reported for non-classfile annotations?

paulp replied:
Can you give me a concrete example of what difference it makes? If someone figures out a way to get the wrong tree here it will throw a nice match error identifying the tree. A match error is a nice concise way to say "unexpected tree".

xeno-by replied:
@lrytz, maybe you could comment on why we have the `UnexpectedTreeAnnotation` in place?

lrytz replied:
I think `MatchError` is equally good here.

paulp committed:
Make sure typed isn't called with an erroneous tree.

I can't see that it makes any difference, but this is
approximately the way it was before.

xeno-by commented:
lgtm. I especially like the shrinkage of `typedAnnotation` signature

paulp commented:
Great, thanks. Tests passing confirmed manually https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/768/consoleText . The failure is another nice spurious one; I also disabled that test.
